File: src/core/graph.rs
use std::collections::HashMap;
use std::collections::HashSet;
use std::hash::Hash;
use std::hash::Hasher;

use super::cluster::Cluster;
use super::dataset::Dataset;
use super::number::Number;

type ClusterSet<'a, T, D> = HashSet<&'a Cluster<'a, T, D>>;
type EdgeSet<'a, T, D> = HashSet<&'a Edge<'a, T, D>>;
type AdjacencyMap<'a, T, D> = HashMap<&'a Cluster<'a, T, D>, ClusterSet<'a, T, D>>;
type FrontierSizes<'a, T, D> = HashMap<&'a Cluster<'a, T, D>, Vec<usize>>;

/// Two `Cluster`s have an `Edge` between them if they have overlapping volumes.
///
/// In CLAM, all `Edge`s are bi-directional.
#[derive(Debug, Clone)]
pub struct Edge<'a, T: Number, D: Dataset<T, f64>> {
    left: &'a Cluster<'a, T, D>,
    right: &'a Cluster<'a, T, D>,
    distance: f64,
}

impl<'a, T: Number, D: Dataset<T, f64>> PartialEq for Edge<'a, T, D> {
    fn eq(&self, other: &Self) -> bool {
        (self.left == other.left) && (self.right == other.right)
    }
}

/// Two `Edge`s are equal if they connect the same two `Cluster`s.
impl<'a, T: Number, D: Dataset<T, f64>> Eq for Edge<'a, T, D> {}

impl<'a, T: Number, D: Dataset<T, f64>> std::fmt::Display for Edge<'a, T, D> {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(f, "{:} -- {:}", self.left, self.right)
    }
}

impl<'a, T: Number, D: Dataset<T, f64>> Hash for Edge<'a, T, D> {
    fn hash<H: Hasher>(&self, state: &mut H) {
        format!("{self}").hash(state)
    }
}

impl<'a, T: Number, D: Dataset<T, f64>> Edge<'a, T, D> {
    /// Creates a new `Edge` from the given `Cluster`s and the distance between
    /// them.
    ///
    /// It is upon the user to verify that the two `Cluster`s are close enough
    /// to have an edge between them.
    pub fn new(left: &'a Cluster<'a, T, D>, right: &'a Cluster<'a, T, D>, distance: f64) -> Self {
        if left < right {
            Self { left, right, distance }
        } else {
            Self {
                left: right,
                right: left,
                distance,
            }
        }
    }

    /// Whether this edge has the given `Cluster` at one of its ends.
    pub fn contains(&self, c: &Cluster<'a, T, D>) -> bool {
        c == self.left || c == self.right
    }

    /// A 2-slice of the `Cluster`s in this `Edge`.
    pub fn clusters(&self) -> [&Cluster<'a, T, D>; 2] {
        [self.left, self.right]
    }

    /// A reference to the `Cluster` at the `left` end of the `Edge`.
    pub fn left(&self) -> &Cluster<'a, T, D> {
        self.left
    }

    /// A reference to the `Cluster` at the `right` end of the `Edge`.
    pub fn right(&self) -> &Cluster<'a, T, D> {
        self.right
    }

    /// The distance between the two `Cluster`s connected by this `Edge`.
    pub fn distance(&self) -> f64 {
        self.distance
    }

    /// Whether this is an edge from a `Cluster` to itself.
    pub fn is_circular(&self) -> bool {
        self.left == self.right
    }

    /// Returns the neighbor of the given `Cluster` in this `Edge`.
    ///
    /// Err:
    ///
    /// * If `c` is not one of the `Cluster`s connected by this `Edge`.
    pub fn neighbor(&self, c: &Cluster<'a, T, D>) -> Result<&Cluster<'a, T, D>, String> {
        if c == self.left {
            Ok(self.right)
        } else if c == self.right {
            Ok(self.left)
        } else {
            Err(format!("Cluster {c} is not in this edge {self}."))
        }
    }
}

/// A `Graph` represents a collection of `Cluster`s and `Edge`s, i.e.
/// connections between overlapping `Cluster`s.
///
/// TODO: Add more info on what graphs we useful for.
#[derive(Debug, Clone)]
pub struct Graph<'a, T: Number, D: Dataset<T, f64>> {
    clusters: ClusterSet<'a, T, D>,
    edges: EdgeSet<'a, T, D>,
    adjacency_map: AdjacencyMap<'a, T, D>,
    population: usize,
    min_depth: usize,
    max_depth: usize,
    ordered_clusters: Vec<&'a Cluster<'a, T, D>>,
    distance_matrix: Option<Vec<Vec<f64>>>,
    adjacency_matrix: Option<Vec<Vec<bool>>>,
    frontier_sizes: Option<FrontierSizes<'a, T, D>>, // TODO: Bench when replacing with DashMap
}

impl<'a, T: Number, D: Dataset<T, f64>> Graph<'a, T, D> {
    /// Create a new `Graph` from the given `clusters` and `edges`. The easiest
    /// and most efficient way to construct a graph is from methods in
    /// `Manifold`.
    ///
    /// # Arguments:
    ///
    /// * `clusters`: The set of `Cluster`s with which to build the `Graph`.
    /// * `edges`: The set of `Edge`s with which to build the `Graph`.
    pub fn new(clusters: ClusterSet<'a, T, D>, edges: EdgeSet<'a, T, D>) -> Self {
        assert!(!clusters.is_empty());

        let (population, min_depth, max_depth) =
            clusters
                .iter()
                .fold((0, usize::MAX, 0), |(population, min_depth, max_depth), &c| {
                    (
                        population + c.cardinality(),
                        std::cmp::min(min_depth, c.depth()),
                        std::cmp::max(max_depth, c.depth()),
                    )
                });

        let adjacency_map = {
            let mut adjacency_map: AdjacencyMap<T, D> = clusters.iter().map(|&c| (c, HashSet::new())).collect();
            edges.iter().for_each(|&e| {
                adjacency_map.get_mut(e.left()).unwrap().insert(e.right());
                adjacency_map.get_mut(e.right()).unwrap().insert(e.left());
            });
            adjacency_map
        };

        Self {
            ordered_clusters: clusters.iter().copied().collect(),
            clusters,
            edges,
            adjacency_map,
            population,
            min_depth,
            max_depth,
            distance_matrix: None,
            adjacency_matrix: None,
            frontier_sizes: None,
        }
    }

    fn compute_distance_matrix(&self) -> Vec<Vec<f64>> {
        let indices = self
            .ordered_clusters
            .iter()
            .enumerate()
            .map(|(i, &c)| (c, i))
            .collect::<HashMap<_, _>>();
        let mut matrix = vec![vec![0.; self.vertex_cardinality()]; self.vertex_cardinality()];
        self.edges.iter().for_each(|&e| {
            let i = *indices.get(e.left()).unwrap();
            let j = *indices.get(e.right()).unwrap();
            matrix[i][j] = e.distance();
            matrix[j][i] = e.distance();
        });
        matrix
    }

    /// Computes the distance matrix for the `Graph` and stores it as an
    /// internal property.
    pub fn with_distance_matrix(mut self) -> Self {
        self.distance_matrix = Some(self.compute_distance_matrix());
        self
    }

    /// Computes the adjacency matrix for the `Graph` and stores it as an
    /// internal property.
    ///
    /// # Panics:
    ///
    /// * If called before calling `with_distance_matrix`.
    pub fn with_adjacency_matrix(mut self) -> Self {
        self.adjacency_matrix = Some(
            self.distance_matrix()
                .iter()
                .map(|row| row.iter().map(|v| *v != 0.).collect())
                .collect(),
        );
        self
    }

    pub fn with_eccentricities(&'a self) -> Self {
        let frontier_sizes = Some(
            self.clusters
                .iter()
                .map(|&c| (c, self.unchecked_traverse(c).1))
                .collect(),
        );

        Self {
            clusters: self.clusters.clone(),
            edges: self.edges.clone(),
            adjacency_map: self.adjacency_map.clone(),
            population: self.population,
            min_depth: self.min_depth,
            max_depth: self.max_depth,
            ordered_clusters: self.ordered_clusters.clone(),
            distance_matrix: self.distance_matrix.clone(),
            adjacency_matrix: self.adjacency_matrix.clone(),
            frontier_sizes,
        }
    }

    #[allow(clippy::manual_retain)]
    pub fn find_component_clusters(&'a self) -> Vec<ClusterSet<'a, T, D>> {
        let mut components = Vec::new();

        let mut unvisited = self.clusters.clone();
        while !unvisited.is_empty() {
            let &start = unvisited.iter().next().unwrap();
            let (visited, _) = self.unchecked_traverse(start);

            // TODO: bench this using `unvisited.retain(|c| !visited.contains(c))`
            unvisited = unvisited.into_iter().filter(|&c| !visited.contains(c)).collect();

            // TODO: Also grab adjacency map, distance matrix, and adjacency matrix
            components.push(visited);
        }

        components
    }

    pub fn clusters(&self) -> &ClusterSet<'a, T, D> {
        &self.clusters
    }

    pub fn edges(&self) -> &EdgeSet<'a, T, D> {
        &self.edges
    }

    pub fn vertex_cardinality(&self) -> usize {
        self.clusters.len()
    }

    pub fn edge_cardinality(&self) -> usize {
        self.edges.len()
    }

    pub fn population(&self) -> usize {
        self.population
    }

    pub fn min_depth(&self) -> usize {
        self.min_depth
    }

    pub fn max_depth(&self) -> usize {
        self.max_depth
    }

    pub fn depth_range(&self) -> (usize, usize) {
        (self.min_depth, self.max_depth)
    }

    pub fn adjacency_map(&'a self) -> &AdjacencyMap<T, D> {
        &self.adjacency_map
    }

    pub fn ordered_clusters(&self) -> &[&Cluster<'a, T, D>] {
        &self.ordered_clusters
    }

    pub fn distance_matrix(&self) -> &Vec<Vec<f64>> {
        self.distance_matrix
            .as_ref()
            .expect("Please call `with_distance_matrix` on the Graph before using `distance_matrix`.")
    }

    pub fn adjacency_matrix(&self) -> &[Vec<bool>] {
        self.adjacency_matrix
            .as_ref()
            .expect("Please call `with_adjacency_matrix` on the Graph before using `adjacency_matrix`.")
    }

    pub fn diameter(&'a self) -> usize {
        self.clusters
            .iter()
            .map(|&c| self.unchecked_eccentricity(c))
            .max()
            .unwrap()
    }

    fn assert_contains(&self, c: &Cluster<'a, T, D>) -> Result<(), String> {
        if self.clusters.contains(&c) {
            Ok(())
        } else {
            Err(format!("Cluster {c} is not in this graph."))
        }
    }

    pub fn unchecked_vertex_degree(&'a self, c: &Cluster<'a, T, D>) -> usize {
        self.unchecked_neighbors_of(c).len()
    }

    pub fn vertex_degree(&'a self, c: &Cluster<'a, T, D>) -> Result<usize, String> {
        self.assert_contains(c)?;
        Ok(self.unchecked_vertex_degree(c))
    }

    pub fn unchecked_neighbors_of(&'a self, c: &Cluster<'a, T, D>) -> &ClusterSet<T, D> {
        self.adjacency_map.get(c).unwrap()
    }

    pub fn neighbors_of(&'a self, c: &Cluster<'a, T, D>) -> Result<&ClusterSet<T, D>, String> {
        self.assert_contains(c)?;
        Ok(self.unchecked_neighbors_of(c))
    }

    pub fn unchecked_traverse(&'a self, start: &'a Cluster<'a, T, D>) -> (ClusterSet<T, D>, Vec<usize>) {
        let mut visited: HashSet<&Cluster<T, D>> = HashSet::new();
        let mut frontier: HashSet<&Cluster<T, D>> = HashSet::new();
        frontier.insert(start);
        let mut frontier_sizes: Vec<usize> = Vec::new();

        while !frontier.is_empty() {
            visited.extend(frontier.iter().copied());
            frontier = frontier
                .iter()
                .flat_map(|&c| self.unchecked_neighbors_of(c))
                .filter(|&n| !((visited.contains(n)) || (frontier.contains(n))))
                .copied()
                .collect();
            frontier_sizes.push(frontier.len());
        }

        (visited, frontier_sizes)
    }

    #[allow(clippy::type_complexity)]
    pub fn traverse(&'a self, start: &'a Cluster<'a, T, D>) -> Result<(ClusterSet<T, D>, Vec<usize>), String> {
        self.assert_contains(start)?;
        Ok(self.unchecked_traverse(start))
    }

    pub fn unchecked_frontier_sizes(&'a self, c: &'a Cluster<'a, T, D>) -> &[usize] {
        self.frontier_sizes
            .as_ref()
            .expect("Please call `with_eccentricities` before using this method.")
            .get(c)
            .unwrap()
    }

    pub fn frontier_sizes(&'a self, c: &'a Cluster<'a, T, D>) -> Result<&[usize], String> {
        self.assert_contains(c)?;
        Ok(self.unchecked_frontier_sizes(c))
    }

    pub fn unchecked_eccentricity(&'a self, c: &'a Cluster<'a, T, D>) -> usize {
        self.unchecked_frontier_sizes(c).len()
    }

    pub fn eccentricity(&'a self, c: &'a Cluster<'a, T, D>) -> Result<usize, String> {
        self.assert_contains(c)?;
        Ok(self.unchecked_eccentricity(c))
    }
}

File: pyclam/core/graph.py
import typing

import numpy

from . import cluster
from . import space
from .. import utils
from ..utils import constants
from ..utils import helpers

logger = helpers.make_logger(__name__)

AdjacencyDict = dict[cluster.Cluster, set['Edge']]


class Edge:
    """ An `Edge` represents a relationship between two `Cluster`s in a `Graph`.
     Two clusters in a graph have an edge between them if there is any overlap
     between the volumes of the two clusters.

    In CLAM, we do not allow edges from a cluster to itself.
    In CLAM, edges are bidirectional.
    """

    __slots__ = [
        '__left',
        '__right',
        '__distance',
    ]

    def __init__(self, left: cluster.Cluster, right: cluster.Cluster, distance: float):
        """ Create an edge using two clusters and the distance between their
         centers.
        """
        self.__left, self.__right = (left, right) if left < right else (right, left)
        self.__distance = distance

    def __eq__(self, other: 'Edge') -> bool:
        """ Two edges are equal if their clusters are equal.
        """
        return self.clusters == other.clusters

    def __str__(self) -> str:
        return f'{str(self.__left)} -- {str(self.__right)}'

    def __repr__(self) -> str:
        return f'{str(self.__left)} -- {str(self.__right)} -- {self.__distance:.6e}'

    def __hash__(self) -> int:
        return hash(str(self))

    def __contains__(self, c: cluster.Cluster) -> bool:
        return (c == self.__left) or (c == self.__right)

    @property
    def clusters(self) -> tuple[cluster.Cluster, cluster.Cluster]:
        return self.__left, self.__right

    @property
    def left(self) -> cluster.Cluster:
        return self.__left

    @property
    def right(self) -> cluster.Cluster:
        return self.__right

    @property
    def distance(self) -> float:
        return self.__distance

    @property
    def to_self(self) -> bool:
        """ Whether the `left` and `right` clusters are equal.
        """
        return self.left == self.right

    def neighbor(self, c: cluster.Cluster) -> cluster.Cluster:
        """ Returns the neighbor of the given cluster, if the cluster is in 
        the edge. Otherwise raises a ValueError.
        """

        if c == self.__right:
            return self.__left

        elif c == self.__left:
            return self.__right

        else:
            raise ValueError(f'Cluster {c.name} is not in this edge {str(self)}.')


class Graph:
    """ A `Graph` is a collection of `Cluster`s with `Edge`s between clusters
    with overlapping volumes. We can select clusters from a tree to build a
    graph.

    All clusters in a graph must come from the same tree and the same
    `MetricSpace`. If using any of the `SelectionCriteria` from CLAM, this
    property will hold. Otherwise, it is upon the user to ensure.

    In CLAM, a `Graph` has a useful invariant:
        - Every instance in the dataset is in exactly one cluster in the graph.
        - No cluster is an ancestor/descendent of another cluster in the graph.

    If clusters were selected using any of the `SelectionCriteria`, this
    invariant will hold. Otherwise, the user must verify this manually.
    This invariant does not hold for sub-graphs such as connected components.
    """

    __slots__ = [
        '__clusters',
        '__root',
        '__metric_space',
        '__population',
        '__min_depth',
        '__max_depth',
        '__edges',
        '__adjacency_dict',
        '__neighborhood_sizes',
        '__components',
        '__distance_matrix',
    ]

    def __init__(self, clusters: set[cluster.Cluster], edges: typing.Optional[set[Edge]] = None):
        """ Creates a graph from the given clusters. The edges need not be
        provided because we have efficient internal algorithms for finding all
        edges.

        Args:
            clusters: The set of clusters with which to build the graph.
            edges: Optional. The set of edges to use.
        """
        if len(clusters) == 0:
            raise ValueError(f'Cannot create a graph with an empty set of clusters.')

        self.__clusters = clusters

        _c = next(iter(clusters))
        self.__root = _c if _c.depth == 0 else _c.ancestry[0]
        self.__metric_space = _c.metric_space

        self.__population = sum(c.cardinality for c in self.__clusters)
        self.__min_depth = min(c.depth for c in self.__clusters)
        self.__max_depth = max(c.depth for c in self.__clusters)

        self.__edges: typing.Union[set[Edge], utils.Unset] = edges or constants.UNSET
        self.__adjacency_dict: typing.Union[AdjacencyDict, utils.Unset] = constants.UNSET
        self.__neighborhood_sizes: dict[cluster.Cluster, list[int]] = dict()
        self.__components: typing.Union[list['Graph'], utils.Unset] = constants.UNSET
        self.__distance_matrix: typing.Union[numpy.ndarray, utils.Unset] = constants.UNSET

    @property
    def root(self) -> cluster.Cluster:
        return self.__root

    @property
    def clusters(self) -> set[cluster.Cluster]:
        return self.__clusters

    @property
    def edges(self) -> set[Edge]:
        if self.__edges is constants.UNSET:
            raise ValueError(f'Please call `build` before accessing this property.')
        return self.__edges

    @property
    def metric_space(self) -> space.Space:
        return self.__metric_space

    @property
    def vertex_cardinality(self) -> int:
        return len(self.__clusters)

    @property
    def edge_cardinality(self) -> int:
        return len(self.edges)

    @property
    def population(self) -> int:
        """ The total number of instances across all clusters in the graph.
        """
        return self.__population

    @property
    def min_depth(self) -> int:
        """ The minimum depth of any cluster in the graph.
        """
        return self.__min_depth

    @property
    def max_depth(self) -> int:
        """ The maximum depth of any cluster in the graph.
        """
        return self.__max_depth

    @property
    def depth_range(self) -> tuple[int, int]:
        """ The range of depths of the clusters in the graph.
        """
        return self.__min_depth, self.__max_depth

    @property
    def adjacency_dict(self) -> AdjacencyDict:
        """ A dictionary of clusters and their sets of edges.
        """
        if self.__adjacency_dict is constants.UNSET:
            raise ValueError(f'Please call `build` before accessing this property.')
        return self.__adjacency_dict

    @property
    def indices(self) -> list[int]:
        """ Indices of instances in all clusters in the graph.
        """
        return [i for c in self.__clusters for i in c.indices]

    @property
    def diameter(self) -> int:
        """ The maximum eccentricity of any cluster in the graph.
        """
        return max(self.eccentricity(c) for c in self.__clusters)

    @property
    def components(self) -> list['Graph']:
        """ List of connected components, as sub-`Graph`s of this graph. If
        there is more than one component, no component will uphold the graph
        invariant.
        """

        if self.__components is constants.UNSET:

            self.__components = list()
            unvisited = self.__clusters.copy()

            while len(unvisited) > 0:
                start = unvisited.pop()
                visited, _ = self.__traverse(start)
                unvisited = unvisited - visited
                self.__components.append(self.subgraph_with(visited))

        return self.__components

    @property
    def distance_matrix(self) -> tuple[list[cluster.Cluster], numpy.ndarray]:
        """ A list of clusters in the graph and a square matrix of edge lengths.
        """

        if self.__distance_matrix is constants.UNSET:

            clusters = list(self.__clusters)
            indices: dict[cluster.Cluster, int] = {c: i for i, c in enumerate(clusters)}

            matrix = numpy.zeros(shape=(self.vertex_cardinality, self.vertex_cardinality))
            for e in self.__edges:
                i, j = indices[e.left], indices[e.right]
                matrix[i, j] = e.distance
                matrix[j, i] = e.distance

            self.__distance_matrix = matrix.astype(numpy.float32)

        return list(self.__clusters), self.__distance_matrix

    @property
    def adjacency_matrix(self) -> tuple[list[cluster.Cluster], numpy.ndarray]:
        """ A list of clusters in the graph and a binary square matrix of edges.
        """
        clusters, matrix = self.distance_matrix
        matrix = (matrix > 0).astype(numpy.uint8)
        return clusters, matrix

    def build(self) -> 'Graph':
        """ Builds the `edges` and `adjacency_dict` of the graph, sets the
        properties and returns the modified graph.
        """
        logger.info(f'Building graph with {self.vertex_cardinality} clusters in a depth range of {self.depth_range} ...')

        if self.__edges is constants.UNSET:
            self.__edges: set[Edge] = {
                Edge(c, n, d) for c in self.__clusters
                for n, d in c.candidate_neighbors.items()
                if (n != c) and (n in self.__clusters) and (d <= (c.radius + n.radius))
            }

        self.__adjacency_dict: AdjacencyDict = {c: set() for c in self.__clusters}
        for e in self.__edges:
            self.__adjacency_dict[e.left].add(e)
            self.__adjacency_dict[e.right].add(e)

        return self

    def assert_contains(self, c: cluster.Cluster):
        """ Raises a ValueError if `c` is not in the graph.
        """
        if c not in self.__clusters:
            raise AssertionError(f'Cluster {c} is not in this Graph.')
        return

    def jaccard(self, other: 'Graph') -> float:
        """ Compute the Jaccard similarity with another graph.
        """
        self_indices = set(self.indices)
        other_indices = set(other.indices)

        intersection = self_indices.intersection(other_indices)
        union = self_indices.union(other_indices)

        return len(intersection) / len(union)

    def vertex_degree(self, c: cluster.Cluster) -> int:
        self.assert_contains(c)
        return len(self.adjacency_dict[c])

    def edges_of(self, c: cluster.Cluster) -> list[Edge]:
        self.assert_contains(c)
        return list(self.adjacency_dict[c])

    def neighbors_of(self, c: cluster.Cluster) -> list[cluster.Cluster]:
        self.assert_contains(c)
        return [e.neighbor(c) for e in self.adjacency_dict[c]]

    def edge_distances_of(self, c: cluster.Cluster) -> list[float]:
        self.assert_contains(c)
        return [e.distance for e in self.adjacency_dict[c]]

    def __traverse(self, start: cluster.Cluster) -> tuple[set[cluster.Cluster], list[int]]:
        """ Performs a traversal, in arbitrary order, starting at the given
        cluster. The traversal continues until no new clusters can be visited.

        Args:
            start: where the traversal will start.

        Returns:
            A 2-tuple of:
                - set of visited clusters.
                - list of number of clusters in the frontier at each iteration
                  of the traversal.
        """

        visited: set[cluster.Cluster] = set()
        frontier: set[cluster.Cluster] = {start}
        frontier_sizes = list()

        while len(frontier) > 0:
            new_frontier = {
                n for c in frontier
                for n in self.neighbors_of(c)
                if not ((n in visited) or (n in frontier))
            }
            visited.update(frontier)
            frontier = new_frontier
            frontier_sizes.append(len(frontier))

        return visited, frontier_sizes

    def frontier_sizes(self, c: cluster.Cluster) -> list[int]:
        self.assert_contains(c)
        if c not in self.__neighborhood_sizes:
            _, n = self.__traverse(c)
            self.__neighborhood_sizes[c] = n
        return self.__neighborhood_sizes[c]

    def eccentricity(self, c: cluster.Cluster) -> int:
        return len(self.frontier_sizes(c))

    def subgraph_with(self, clusters: set[cluster.Cluster]) -> 'Graph':
        [self.assert_contains(c) for c in clusters]
        edges = {
            e for e in self.__edges
            if (e.left in clusters) and (e.right in clusters)
        }
        return Graph(clusters, edges).build()

    def component_containing(self, c: cluster.Cluster) -> 'Graph':
        self.assert_contains(c)
        for component in self.components:
            if c in component.__clusters:
                return component
        else:
            raise ValueError(f'Cluster is the graph but not in any component. This is a bug.')

    def as_dot_string(self) -> str:
        raise NotImplementedError

    @staticmethod
    def from_dot_string(dot_string) -> 'Graph':
        raise NotImplementedError

    def __add(self, c: cluster.Cluster):
        raise NotImplementedError

    def __remove(self, c: cluster.Cluster):
        raise NotImplementedError

    def replace_clusters(self, removals: set[cluster.Cluster], additions: set[cluster.Cluster]) -> 'Graph':
        raise NotImplementedError


__all__ = [
    'AdjacencyDict',
    'Edge',
    'Graph',
]
